### PR TITLE
#12 Solution pan gesture

### PR DESCRIPTION
**Pull request by kanottonp**
A solution for pan gesture without using AssociationPolicy

### CODE REVIEW

File: Sources/SwipePopViewController/SwipePopViewController.swift [hunk 1]
```diff
@@ -7,147 +7,90 @@
...
 
-fileprivate enum ViewAssociatedKeys {
-    static var firstTranslation = "swipepop.first_translation"
-    static var panGestureRecognizer = "swipepop.pan_gesture_recognizer"
-    static var selfNavigationControllerDelegate = "swipepop.self_navigation_controller_delegate"
-    static var percentDrivenInteractiveTransition = "swipepop.percent_drivenInteractive_transition"
-    static var panGestureInteractor = "swipepop.pan_gesture_interactor"
-    static var swipePopNavigationControllerDelegate = "swipepop.swipe_pop_navigation_controller_delegate"
-}
+class PanInteractivePopGesture: UIPanGestureRecognizer {
```
Comment (chalermpong) [inline Sources/SwipePopViewController/SwipePopViewController.swift:11]: Put this class in a new file

File: Sources/SwipePopViewController/SwipePopViewController.swift [hunk 2]
```diff
@@ -7,147 +7,90 @@
...
     }
     
-    var panGestureRecognizer: UIPanGestureRecognizer? {
-        get { return ao_get(pkey: &ViewAssociatedKeys.panGestureRecognizer) as? UIPanGestureRecognizer }
-        set { ao_setOptional(newValue, pkey: &ViewAssociatedKeys.panGestureRecognizer) }
+    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent) {
+        print("touchesBegan")
+        super.touchesBegan(touches, with: event)
+        
     }
```
Comment (chalermpong) [inline Sources/SwipePopViewController/SwipePopViewController.swift:30]: Remove all the function that only prints debug message

File: Sources/SwipePopViewController/SwipePopViewController.swift [hunk 3]
```diff
@@ -7,147 +7,90 @@
...
+        super.touchesCancelled(touches, with: event)
+        firstTranslation = .zero
     }
     
-    @objc func handlePanGesture(_ panGesture: UIPanGestureRecognizer) {
-        guard let view = self.view else { return }
-        let percent = self.firstTranslation != nil ? max(panGesture.translation(in: view).x - self.firstTranslation!, 0) / view.frame.width : 0
-        
-        switch panGesture.state {
+    var firstTranslation: CGPoint = .zero
```
Comment (chalermpong) [inline Sources/SwipePopViewController/SwipePopViewController.swift:52]: Move this variable to beginning of this class

File: Sources/SwipePopViewController/SwipePopViewController.swift [hunk 4]
```diff
@@ -7,147 +7,90 @@
...
+    var lastBeginTime = Date()
```
Comment (chalermpong) [inline Sources/SwipePopViewController/SwipePopViewController.swift:53]: Remove this variable (it's just for debugging)

File: Sources/SwipePopViewController/SwipePopViewController.swift [hunk 5]
```diff
@@ -7,147 +7,90 @@
...
 
-class SwipePopNavigationControllerDelegate: NSObject, UINavigationControllerDelegate {
-    weak var viewControllerDelegate: UIViewController?
-    
-    init(viewControllerDelegate: UIViewController?) {
-        self.viewControllerDelegate = viewControllerDelegate
+    override func velocity(in view: UIView?) -> CGPoint {
+        let v = super.velocity(in: view)
+        print("velocity \(v) at \(Date().timeIntervalSince(lastBeginTime))")
+        return v
```
Comment (chalermpong) [inline Sources/SwipePopViewController/SwipePopViewController.swift:77]: This function can be removed

File: Sources/SwipePopViewController/SwipePopViewController.swift [hunk 6]
```diff
@@ -7,147 +7,90 @@
...
     }
+
     
-    func navigationController(_ navigationController: UINavigationController, animationControllerFor operation: UINavigationController.Operation, from fromVC: UIViewController, to toVC: UIViewController) -> UIViewControllerAnimatedTransitioning? {
-        return SlideAnimatedTransitioning()
+    deinit {
+        print("deinit")
```
Comment (chalermpong) [inline Sources/SwipePopViewController/SwipePopViewController.swift:82]: Remove

File: Sources/SwipePopViewController/SwipePopViewController.swift [hunk 7]
```diff
@@ -7,147 +7,90 @@
...
-    static var swipePopNavigationControllerDelegate = "swipepop.swipe_pop_navigation_controller_delegate"
-}
+class PanInteractivePopGesture: UIPanGestureRecognizer {
 
-extension UIViewController {
-    var firstTranslation: CGFloat? {
-        get { return ao_get(pkey: &ViewAssociatedKeys.firstTranslation) as? CGFloat }
-        set { ao_setOptional(newValue, pkey: &ViewAssociatedKeys.firstTranslation) }
+    func addInteractivePopGesture(navigationController: UINavigationController,
+                                  view: UIView) {
```
Comment (chalermpong) [inline Sources/SwipePopViewController/SwipePopViewController.swift:14]: Move this logic to `UIViewController.addSwipePopGesture`

File: Example/SwipePopViewController/ViewController.swift [hunk 8]
```diff
@@ -31,7 +31,7 @@ class ViewController: UIViewController {
         
         switch mode {
         case .native:
-            applyNativePopGesture()
+            addSwipePopGesture()
```
Comment (chalermpong) [inline Example/SwipePopViewController/ViewController.swift:34]: Please restore the old native code (`applyNativePopGesture` should not be changed)

**Review by chalermpong — CHANGES_REQUESTED** [review]

**Review by chalermpong — APPROVED** [review]

**Comment by chalermpong** [comment]
PR for #12

**Review by camp191 — APPROVED** [review]

**Review by chalermpong — APPROVED** [review]

**Pull request merged**